language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala

```diff
@@ -18,67 +17,94 @@ object ClassForType extends StrictLogging {
-          className.simpleName().toLowerCase
-        )
-        val (tpe, constructors) =
+        javaFiles(packageName, recordClass)
+      case Normal(DefDataType(typeVars, variant: Variant.FWT)) =>
+        val simpleLowerCaseName = JavaEscaper.escapeString(className.simpleName().toLowerCase)
+        val subPackage = s"$packageName.$simpleLowerCaseName"
+        val escapedTypeVars = typeVars.map(JavaEscaper.escapeString)
+        val (variantSpec, constructorSpecs) =
           VariantClass.generate(
             className,
             subPackage,
-            typeVars.map(JavaEscaper.escapeString),
+            escapedTypeVars,
             variant,
             typeWithContext,
             packagePrefixes,
           )
-        javaFile(typeWithContext, javaPackage, tpe) ::
-          constructors.map(cons => javaFile(typeWithContext, subPackage, cons))
-
-      case Some(Normal(DefDataType(_, enum: Enum))) =>
-        List(
-          JavaFile
-            .builder(javaPackage, EnumClass.generate(className, enum))
-            .build()
-        )
-
-      case Some(Template(record, template)) =>
+        javaFile(packageName, variantSpec) :: javaFiles(subPackage, constructorSpecs)
+      case Normal(DefDataType(_, enum: Enum)) =>
+        javaFiles(packageName, EnumClass.generate(className, enum))
+      case Template(record, template) =>
         val typeSpec =
-          TemplateClass.generate(className, record, template, typeWithContext, packagePrefixes)
-        List(JavaFile.builder(javaPackage, typeSpec).build())
-
-      case None =>
-        // This typeWithContext didn't contain a type itself, but has children nodes
-        // which we treat as any other TypeWithContext
-        typeWithContext.typesLineages.flatMap(ClassForType(_, packagePrefixes)).toList
+          TemplateClass.generate(
+            className,
+            record,
+            template,
+            typeWithContext,
+            packagePrefixes,
+          )
+        javaFiles(packageName, typeSpec)
     }
   }
 
-  def javaFile(typeWithContext: TypeWithContext, javaPackage: String, typeSpec: TypeSpec) = {
-    val withField =
-      typeSpec.toBuilder.addField(createPackageIdField(typeWithContext.interface.packageId)).build()
-    JavaFile.builder(javaPackage, withField).build()
-  }
+  private def javaFile(packageName: String, typeSpec: TypeSpec): JavaFile =
+    JavaFile.builder(packageName, typeSpec).build()
+
+  private def javaFiles(packageName: String, typeSpec: TypeSpec): List[JavaFile] =
+    List(javaFile(packageName, typeSpec))
+
+  private def javaFiles(packageName: String, typeSpecs: List[TypeSpec]): List[JavaFile] =
+    typeSpecs.map(javaFile(packageName, _))
 
-  private def createPackageIdField(packageId: PackageId): FieldSpec = {
-    FieldSpec
-      .builder(classOf[String], "_packageId", Modifier.FINAL, Modifier.PUBLIC, Modifier.STATIC)
-      .initializer("$S", packageId)
-      .build()
-  }
```

Comment:
This needed to be generated specifically for templates and variants. Instead of having the logic here I moved it to the place where the types are originally created. The logic is still centralized in a single method (which I moved to the package object).